### PR TITLE
Retry HTTP requests that are temporarily failing

### LINKED ISSUE #1556: Pub should handle pub server errors gracefully
The pub server will sometimes return a 500 (internal error), if there is a temporary failure on the server.
pub get should handle these results, and retry the fetch of the object from the server.  If it doesn't retry, then it should return with a distinctive error code that indicates temporary failure.  Maybe this should happen for other network failures when running pub get as well.

This would make the testing of pub get less flaky, and allow us to test it on the buildbot.
This seems related to issue #1534 and pr https://github.com/dart-lang/pub/pull/1552

@mkustermann

### PR DESCRIPTION
**Pull request by nex3**
Closes #1556

### CODE REVIEW

File: lib/src/http.dart [hunk 1]
```diff
@@ -210,8 +212,36 @@ class _PubHttpClient extends http.BaseClient {
...
+            // where an AppEngine instance will go down and need to be rebooted,
+            // which takes about a minute.
+            return new Duration(milliseconds: 30);
+          }
+        },
+        onRetry: (request, response, retryCount) {
+          log.io("Retry #${retryCount + 1} for "
+              "${request.method} ${request.url}...");
+          if (retryCount != 3) return;
+          if (!_retryMessages.add(request.url.host)) return;
```
Comment (kevmoo) [inline lib/src/http.dart]: What does `_retryMessages` do?

File: lib/src/http.dart [hunk 2]
```diff
@@ -210,8 +212,36 @@ class _PubHttpClient extends http.BaseClient {
...
+        when: (response) => const [502, 503, 504].contains(response.statusCode),
+        delay: (retryCount) {
+          if (retryCount < 3) {
+            // Retry quickly a couple times in case of a short transient error.
+            return new Duration(milliseconds: 500) * math.pow(1.5, retryCount);
+          } else {
+            // If the error persists, wait a long time. This works around issues
+            // where an AppEngine instance will go down and need to be rebooted,
+            // which takes about a minute.
+            return new Duration(milliseconds: 30);
```
Comment (kevmoo) [inline lib/src/http.dart]: 30ms is a long time? 😄

**Comment by kevmoo** [comment]
Also see travis...

File: lib/src/http.dart (hunk 2 above)
Comment (nex3) [reply lib/src/http.dart]: Oops, fixed.

File: lib/src/http.dart (hunk 1 above)
Comment (nex3) [reply lib/src/http.dart]: As the documentation says, it tracks which hostnames we've already printed messages about.

Comment (kevmoo) [reply lib/src/http.dart]: Perhaps `_retryHosts` then?

To understand this: on the 3rd retry for a given request (when we start getting slow), on the first time we're having issues w/ this host, log the friendly message.

Right?

File: lib/src/http.dart [hunk 3]
```diff
@@ -210,8 +212,36 @@ class _PubHttpClient extends http.BaseClient {
 /// The [_PubHttpClient] wrapped by [httpClient].
 final _pubClient = new _PubHttpClient();
 
+/// A set of all hostnames for which we've printed a message indicating that
+/// we're waiting for them to come back up.
+final _retryMessages = new Set<String>();
```
Comment (jakemac53) [inline lib/src/http.dart]: nit: consider `retriedHosts` or something along those lines - the name doesn't seem to match the description particularly well.

File: lib/src/http.dart [hunk 4]
```diff
@@ -210,8 +212,36 @@ class _PubHttpClient extends http.BaseClient {
...
-final httpClient = new ThrottleClient(16, _pubClient);
+final httpClient = new ThrottleClient(
+    16,
+    new RetryClient(_pubClient,
+        retries: 5,
+        when: (response) => const [502, 503, 504].contains(response.statusCode),
+        delay: (retryCount) {
+          if (retryCount < 3) {
+            // Retry quickly a couple times in case of a short transient error.
+            return new Duration(milliseconds: 500) * math.pow(1.5, retryCount);
```
Comment (jakemac53) [inline lib/src/http.dart]: It is a good idea to also add in an additional random delay - you would keep exactly this logic but just add between maybe 0 and 500 additional milliseconds. This helps to ensure that when a lot of requests are sent at once they don't all keep rescheduling on the same intervals.

File: lib/src/http.dart (hunk 3 above)
Comment (nex3) [reply lib/src/http.dart]: Done.

File: lib/src/http.dart (hunk 4 above)
Comment (nex3) [reply lib/src/http.dart]: Done.

File: lib/src/http.dart (hunk 1 above)
Comment (nex3) [reply lib/src/http.dart]: > Perhaps `_retryHosts` then?

Done.

> To understand this: on the 3rd retry for a given request (when we start getting slow), on the first time we're having issues w/ this host, log the friendly message.

That's right.

**Comment by kevmoo** [comment]
Land it – can't wait to get this into the SDK

**Pull request merged**